src/main/java/com/microsoft/sqlserver/jdbc/SQLServerDataTable.java

```diff
@@ -207,92 +217,86 @@ private void internalAddrow(JDBCType jdbcType, Object val, Object[] rowValues,
-                    val = val.toString();
-            case VARCHAR:
-            case NCHAR:
-            case NVARCHAR:
-            case LONGVARCHAR:
-            case LONGNVARCHAR:
-            case SQLXML:
-                bValueNull = (null == val);
-                nValueLen = bValueNull ? 0 : (2 * ((String) val).length());
-
-                if (nValueLen > currentColumnMetadata.precision) {
-                    currentColumnMetadata.precision = nValueLen;
-                    columnMetadata.put(pair.getKey(), currentColumnMetadata);
-                }
-                rowValues[pair.getKey()] = (bValueNull) ? null : (String) val;
-                break;
-            case SQL_VARIANT:
-                JDBCType internalJDBCType;
-                if (null == val) { // TODO:Check this later
-                    throw new SQLServerException(SQLServerException.getErrString("R_invalidValueForTVPWithSQLVariant"),
-                            null);
-                }
-                JavaType javaType = JavaType.of(val);
-                internalJDBCType = javaType.getJDBCType(SSType.UNKNOWN, jdbcType);
-                internalAddrow(internalJDBCType, val, rowValues, pair);
-                break;
-            default:
-                MessageFormat form = new MessageFormat(SQLServerException.getErrString("R_unsupportedDataTypeTVP"));
-                Object[] msgArgs = {jdbcType};
-                throw new SQLServerException(null, form.format(msgArgs), null, 0, false);
+                    rowValues[pair.getKey()] = bd;
+                    break;
+
+                case DOUBLE:
+                    rowValues[pair.getKey()] = (val instanceof Double) ? val : Double.parseDouble(val.toString());
+                    break;
+
+                case FLOAT:
+                case REAL:
+                    rowValues[pair.getKey()] = (val instanceof Float) ? val : Float.parseFloat(val.toString());
+                    break;
+
+                case TIMESTAMP_WITH_TIMEZONE:
+                case TIME_WITH_TIMEZONE:
+                case DATE:
+                case TIME:
+                case TIMESTAMP:
+                case DATETIMEOFFSET:
+                case DATETIME:
+                case SMALLDATETIME:
+                    // Sending temporal types as string. Error from database is thrown if parsing fails
+                    // no need to send precision for temporal types, string literal will never exceed
+                    // DataTypes.SHORT_VARTYPE_MAX_BYTES
+
+                    // java.sql.Date, java.sql.Time and java.sql.Timestamp are subclass of java.util.Date
+                    if (val instanceof java.util.Date || val instanceof microsoft.sql.DateTimeOffset
+                            || val instanceof OffsetDateTime || val instanceof OffsetTime)
+                        rowValues[pair.getKey()] = val.toString();
+                    else
+                        rowValues[pair.getKey()] = val;
```

Comment:
Don't we need to cast this to String?

Reply:
Changes done as per suggestion here: https://github.com/Microsoft/mssql-jdbc/pull/990#discussion_r268369804
Casting is not required as the object type is expected to be of required type. If it is not, it wouldn't get cast eitherway.

Reply:
That comment is addressing a different scenario. In that scenario, we've already confirmed that `val` is a Long instance, so there is no further casting required, and we were unnecessarily unboxing the value. Here, we haven't confirmed the type of `val`, and we need to send the temporal types as String, as mentioned in the comments.

Reply:
Temporal types are already covered in the `if` block and their String value is determined, `else` case is for handling String type hence it was cast to String which is not necessary.